src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Common.Core.Diagnostics;
+
+namespace Microsoft.Common.Core.Services {
+    public class ServiceManager : IServiceManager {
+        class ServiceFactory {
+            public IServiceFactory Factory { get; set; }
+            public object[] Parameters { get; set; }
+        }
+
+        private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
+        private readonly Dictionary<Type, ServiceFactory> _deferredServices = new Dictionary<Type, ServiceFactory>();
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Fire when service is added
+        /// </summary>
+        public event EventHandler<ServiceContainerEventArgs> ServiceAdded;
+        /// <summary>
+        /// Fires when service is removed
+        /// </summary>
+        public event EventHandler<ServiceContainerEventArgs> ServiceRemoved;
+
+        /// <summary>
+        /// Add service to the service manager container
+        /// </summary>
+        /// <typeparam name="T">Service type</typeparam>
+        /// <param name="service">Service instance</param>
+        public virtual IServiceContainer AddService<T>(T service) where T : class {
+            var type = typeof(T);
+            lock (_lock) {
+                Check.ArgumentNull(nameof(service), service);
+                Check.InvalidOperation(() => _services.ContainsKey(type));
+
+                _services[type] = service;
+                ServiceAdded?.Invoke(this, new ServiceContainerEventArgs(type));
```

Comment:
It is usually a bad idea to raise event under lock - subscriber can block the whole class.